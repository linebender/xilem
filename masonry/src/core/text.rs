// Copyright 2018 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

//! Support for text display and rendering
//!
//! There are three kinds of text commonly needed:
//!  1) Non interactive text (e.g. a button's label)
//!  2) Selectable text (e.g. a paragraph of content)
//!  3) Editable text (e.g. a search bar)
//!
//! All of these have the same set of global styling options, and can contain rich text

use parley::GenericFamily;

/// A reference counted string slice.
///
/// This is a data-friendly way to represent strings in Masonry. Unlike `String`
/// it cannot be mutated, but unlike `String` it can be cheaply cloned.
pub type ArcStr = std::sync::Arc<str>;

/// The Parley [`Brush`] used within Masonry.
///
/// This enables updating of brush details without performing relayouts;
/// the inner values are indexes into the `brushes` argument to [`render_text()`].
///
/// [`Brush`]: crate::parley::Brush
#[derive(Clone, PartialEq, Default, Debug)]
pub struct BrushIndex(pub usize);

/// A style property specialised for use within Masonry.
pub type StyleProperty = parley::StyleProperty<'static, BrushIndex>;

/// A set of styles specialised for use within Masonry.
pub type StyleSet = parley::StyleSet<BrushIndex>;

/// Applies the default text styles for Masonry into `styles`.
pub(crate) fn default_styles(styles: &mut StyleSet) {
    styles.insert(StyleProperty::LineHeight(1.2));
    styles.insert(GenericFamily::SystemUi.into());
}

use parley::{Layout, PositionedLayoutItem};
use vello::kurbo::{Affine, Line, Stroke};
use vello::peniko::{Brush, Fill};
use vello::Scene;

/// A function that renders laid out glyphs to a [`Scene`].
///
/// The `BrushIndex` values of the runs are indices into `brushes`.
pub fn render_text(
    scene: &mut Scene,
    transform: Affine,
    layout: &Layout<BrushIndex>,
    brushes: &[Brush],
    // TODO: Should this be part of `BrushIndex` (i.e. `brushes`)?
    hint: bool,
) {
    for line in layout.lines() {
        for item in line.items() {
            let PositionedLayoutItem::GlyphRun(glyph_run) = item else {
                continue;
            };
            let style = glyph_run.style();
            // We draw underlines under the text, then the strikethrough on top, following:
            // https://drafts.csswg.org/css-text-decor/#painting-order
            if let Some(underline) = &style.underline {
                let underline_brush = &brushes[underline.brush.0];
                let run_metrics = glyph_run.run().metrics();
                let offset = match underline.offset {
                    Some(offset) => offset,
                    None => run_metrics.underline_offset,
                };
                let width = match underline.size {
                    Some(size) => size,
                    None => run_metrics.underline_size,
                };
                // The `offset` is the distance from the baseline to the top of the underline
                // so we move the line down by half the width
                // Remember that we are using a y-down coordinate system
                // If there's a custom width, because this is an underline, we want the custom
                // width to go down from the default expectation
                let y = glyph_run.baseline() - offset + width / 2.;

                let line = Line::new(
                    (glyph_run.offset() as f64, y as f64),
                    ((glyph_run.offset() + glyph_run.advance()) as f64, y as f64),
                );
                scene.stroke(
                    &Stroke::new(width.into()),
                    transform,
                    underline_brush,
                    None,
                    &line,
                );
            }
            let mut x = glyph_run.offset();
            let y = glyph_run.baseline();
            let run = glyph_run.run();
            let font = run.font();
            let font_size = run.font_size();
            let synthesis = run.synthesis();
            let glyph_xform = synthesis
                .skew()
                .map(|angle| Affine::skew(angle.to_radians().tan() as f64, 0.0));
            let coords = run.normalized_coords();
            let brush = &brushes[style.brush.0];
            scene
                .draw_glyphs(font)
                .brush(brush)
                .hint(hint)
                .transform(transform)
                .glyph_transform(glyph_xform)
                .font_size(font_size)
                .normalized_coords(coords)
                .draw(
                    Fill::NonZero,
                    glyph_run.glyphs().map(|glyph| {
                        let gx = x + glyph.x;
                        let gy = y - glyph.y;
                        x += glyph.advance;
                        vello::Glyph {
                            id: glyph.id as _,
                            x: gx,
                            y: gy,
                        }
                    }),
                );

            if let Some(strikethrough) = &style.strikethrough {
                let strikethrough_brush = &brushes[strikethrough.brush.0];
                let run_metrics = glyph_run.run().metrics();
                let offset = match strikethrough.offset {
                    Some(offset) => offset,
                    None => run_metrics.strikethrough_offset,
                };
                let width = match strikethrough.size {
                    Some(size) => size,
                    None => run_metrics.strikethrough_size,
                };
                // The `offset` is the distance from the baseline to the *top* of the strikethrough
                // so we calculate the middle y-position of the strikethrough based on the font's
                // standard strikethrough width.
                // Remember that we are using a y-down coordinate system
                let y = glyph_run.baseline() - offset + run_metrics.strikethrough_size / 2.;

                let line = Line::new(
                    (glyph_run.offset() as f64, y as f64),
                    ((glyph_run.offset() + glyph_run.advance()) as f64, y as f64),
                );
                scene.stroke(
                    &Stroke::new(width.into()),
                    transform,
                    strikethrough_brush,
                    None,
                    &line,
                );
            }
        }
    }
}
