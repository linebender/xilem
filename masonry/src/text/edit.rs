// Copyright 2018 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

use std::ops::{Deref, DerefMut, Range};

use kurbo::Point;
use parley::{FontContext, LayoutContext};
use vello::Scene;
use winit::{
    event::Ime,
    keyboard::{Key, NamedKey},
};

use crate::{
    event::{PointerButton, PointerState},
    Action, EventCtx, Handled, TextEvent,
};

use super::{
    offset_for_delete_backwards,
    selection::{Affinity, Selection},
    Selectable, TextBrush, TextWithSelection,
};

/// Text which can be edited
pub trait EditableText: Selectable {
    /// Replace range with new text.
    /// Can panic if supplied an invalid range.
    // TODO: make this generic over Self
    fn edit(&mut self, range: Range<usize>, new: impl Into<String>);
    /// Create a value of this struct
    fn from_str(s: &str) -> Self;
}

impl EditableText for String {
    fn edit(&mut self, range: Range<usize>, new: impl Into<String>) {
        self.replace_range(range, &new.into());
    }
    fn from_str(s: &str) -> Self {
        s.to_string()
    }
}

// TODO: What advantage does this actually have?
// impl EditableText for Arc<String> {
//     fn edit(&mut self, range: Range<usize>, new: impl Into<String>) {
//         let new = new.into();
//         if !range.is_empty() || !new.is_empty() {
//             Arc::make_mut(self).edit(range, new)
//         }
//     }
//     fn from_str(s: &str) -> Self {
//         Arc::new(s.to_owned())
//     }
// }

/// A region of text which can support editing operations
pub struct TextEditor<T: EditableText> {
    inner: TextWithSelection<T>,
    /// The range of the preedit region in the text
    preedit_range: Option<Range<usize>>,
}

impl<T: EditableText> TextEditor<T> {
    pub fn new(text: T, text_size: f32) -> Self {
        Self {
            inner: TextWithSelection::new(text, text_size),
            preedit_range: None,
        }
    }

    pub fn reset_preedit(&mut self) {
        self.preedit_range = None;
    }

    /// Rebuild the text.
    ///
    /// See also [TextLayout::rebuild](crate::text2::TextLayout::rebuild) for more comprehensive docs.
    pub fn rebuild(
        &mut self,
        font_ctx: &mut FontContext,
        layout_ctx: &mut LayoutContext<TextBrush>,
    ) {
        self.inner
            .rebuild_with_attributes(font_ctx, layout_ctx, |mut builder| {
                if let Some(range) = self.preedit_range.as_ref() {
                    builder.push(
                        &parley::style::StyleProperty::Underline(true),
                        range.clone(),
                    );
                }
                builder
            });
    }

    pub fn draw(&mut self, scene: &mut Scene, point: impl Into<Point>) {
        self.inner.draw(scene, point);
    }

    pub fn pointer_down(
        &mut self,
        origin: Point,
        state: &PointerState,
        button: PointerButton,
    ) -> bool {
        // TODO: If we have a selection and we're hovering over it,
        // implement (optional?) click and drag
        self.inner.pointer_down(origin, state, button)
    }

    pub fn text_event(&mut self, ctx: &mut EventCtx, event: &TextEvent) -> Handled {
        let inner_handled = self.inner.text_event(event);
        if inner_handled.is_handled() {
            return inner_handled;
        }
        match event {
            TextEvent::KeyboardKey(event, mods) if event.state.is_pressed() => {
                // We don't input actual text when these keys are pressed
                if !(mods.control_key() || mods.alt_key() || mods.super_key()) {
                    match &event.logical_key {
                        Key::Named(NamedKey::Backspace) => {
                            if let Some(selection) = self.inner.selection {
                                if !selection.is_caret() {
                                    self.text_mut().edit(selection.range(), "");
                                    self.inner.selection =
                                        Some(Selection::caret(selection.min(), Affinity::Upstream));
                                } else {
                                    // TODO: more specific behavior may sometimes be warranted here
                                    //       because whole EGCs are more coarse than what people expect
                                    //       to be able to delete individual indic grapheme cluster
                                    //       components among other things.
                                    let text = self.text_mut();
                                    let offset =
                                        offset_for_delete_backwards(selection.active, text);
                                    self.text_mut().edit(offset..selection.active, "");
                                    self.inner.selection =
                                        Some(Selection::caret(offset, selection.active_affinity));
                                }
                                Handled::Yes
                            } else {
                                Handled::No
                            }
                        }
                        Key::Named(NamedKey::Delete) => {
                            if let Some(selection) = self.inner.selection {
                                if !selection.is_caret() {
                                    self.text_mut().edit(selection.range(), "");
                                    self.inner.selection = Some(Selection::caret(
                                        selection.min(),
                                        Affinity::Downstream,
                                    ));
                                } else if let Some(offset) =
                                    self.text().next_grapheme_offset(selection.active)
                                {
                                    self.text_mut().edit(selection.min()..offset, "");
                                    self.inner.selection = Some(Selection::caret(
                                        selection.min(),
                                        selection.active_affinity,
                                    ));
                                }
                                Handled::Yes
                            } else {
                                Handled::No
                            }
                        }
                        Key::Named(NamedKey::Space) => {
                            let selection = self.inner.selection.unwrap_or(Selection {
                                anchor: 0,
                                active: 0,
                                active_affinity: Affinity::Downstream,
                                h_pos: None,
                            });
                            let c = ' ';
                            self.text_mut().edit(selection.range(), c);
                            self.inner.selection = Some(Selection::caret(
                                selection.min() + c.len_utf8(),
                                // We have just added this character, so we are "affined" with it
                                Affinity::Downstream,
                            ));
                            let contents = self.text().as_str().to_string();
                            ctx.submit_action(Action::TextChanged(contents));
                            Handled::Yes
                        }
                        Key::Named(NamedKey::Enter) => {
                            let contents = self.text().as_str().to_string();
                            ctx.submit_action(Action::TextEntered(contents));
                            Handled::Yes
                        }
                        Key::Named(_) => Handled::No,
                        Key::Character(c) => {
                            let selection = self.inner.selection.unwrap_or(Selection {
                                anchor: 0,
                                active: 0,
                                active_affinity: Affinity::Downstream,
                                h_pos: None,
                            });
                            self.text_mut().edit(selection.range(), &**c);
                            self.inner.selection = Some(Selection::caret(
                                selection.min() + c.len(),
                                // We have just added this character, so we are "affined" with it
                                Affinity::Downstream,
                            ));
                            let contents = self.text().as_str().to_string();
                            ctx.submit_action(Action::TextChanged(contents));
                            Handled::Yes
                        }
                        Key::Unidentified(_) => Handled::No,
                        Key::Dead(d) => {
                            eprintln!("Got dead key {d:?}. Will handle");
                            Handled::No
                        }
                    }
                } else if mods.control_key() || mods.super_key()
                // TODO: do things differently on mac, rather than capturing both super and control.
                {
                    match &event.logical_key {
                        Key::Named(NamedKey::Backspace) => {
                            if let Some(selection) = self.inner.selection {
                                if !selection.is_caret() {
                                    self.text_mut().edit(selection.range(), "");
                                    self.inner.selection =
                                        Some(Selection::caret(selection.min(), Affinity::Upstream));
                                }
                                let offset =
                                    self.text().prev_word_offset(selection.active).unwrap_or(0);
                                self.text_mut().edit(offset..selection.active, "");
                                self.inner.selection =
                                    Some(Selection::caret(offset, Affinity::Upstream));

                                let contents = self.text().as_str().to_string();
                                ctx.submit_action(Action::TextChanged(contents));
                                Handled::Yes
                            } else {
                                Handled::No
                            }
                        }
                        Key::Named(NamedKey::Delete) => {
                            if let Some(selection) = self.inner.selection {
                                if !selection.is_caret() {
                                    self.text_mut().edit(selection.range(), "");
                                    self.inner.selection = Some(Selection::caret(
                                        selection.min(),
                                        Affinity::Downstream,
                                    ));
                                } else if let Some(offset) =
                                    self.text().next_word_offset(selection.active)
                                {
                                    self.text_mut().edit(selection.active..offset, "");
                                    self.inner.selection =
                                        Some(Selection::caret(selection.min(), Affinity::Upstream));
                                }
                                let contents = self.text().as_str().to_string();
                                ctx.submit_action(Action::TextChanged(contents));
                                Handled::Yes
                            } else {
                                Handled::No
                            }
                        }
                        _ => Handled::No,
                    }
                } else {
                    Handled::No
                }
            }
            TextEvent::KeyboardKey(_, _) => Handled::No,
            TextEvent::Ime(ime) => match ime {
                Ime::Commit(text) => {
                    if let Some(selection_range) = self.selection.map(|x| x.range()) {
                        self.text_mut().edit(selection_range.clone(), text);
                        self.selection = Some(Selection::caret(
                            selection_range.start + text.len(),
                            Affinity::Upstream,
                        ));
                    }
                    let contents = self.text().as_str().to_string();
                    ctx.submit_action(Action::TextChanged(contents));
                    Handled::Yes
                }
                Ime::Preedit(preedit_string, preedit_sel) => {
                    if let Some(preedit) = self.preedit_range.clone() {
                        // TODO: Handle the case where this is the same value, to avoid some potential infinite loops
                        self.text_mut().edit(preedit.clone(), preedit_string);
                        let np = preedit.start..(preedit.start + preedit_string.len());
                        self.preedit_range = if preedit_string.is_empty() {
                            None
                        } else {
                            Some(np.clone())
                        };
                        self.selection = if let Some(pec) = preedit_sel {
                            Some(Selection::new(
                                np.start + pec.0,
                                np.start + pec.1,
                                Affinity::Upstream,
                            ))
                        } else {
                            Some(Selection::caret(np.end, Affinity::Upstream))
                        };
                    } else {
                        // If we've been sent an event to clear the preedit,
                        // but there was no existing pre-edit, there's nothing to do
                        // so we report that the event has been handled
                        // An empty preedit is sent by some environments when the
                        // context of a text input has changed, even if the contents
                        // haven't; this also avoids some potential infinite loops
                        if preedit_string.is_empty() {
                            return Handled::Yes;
                        }
                        let sr = self.selection.map(|x| x.range()).unwrap_or(0..0);
                        self.text_mut().edit(sr.clone(), preedit_string);
                        let np = sr.start..(sr.start + preedit_string.len());
                        self.preedit_range = if preedit_string.is_empty() {
                            None
                        } else {
                            Some(np.clone())
                        };
                        self.selection = if let Some(pec) = preedit_sel {
                            Some(Selection::new(
                                np.start + pec.0,
                                np.start + pec.1,
                                Affinity::Upstream,
                            ))
                        } else {
                            Some(Selection::caret(np.start, Affinity::Upstream))
                        };
                    }
                    Handled::Yes
                }
                Ime::Enabled => {
                    // Generally this shouldn't happen, but I can't prove it won't.
                    if let Some(preedit) = self.preedit_range.clone() {
                        self.text_mut().edit(preedit.clone(), "");
                        self.selection = Some(
                            self.selection
                                .unwrap_or(Selection::caret(0, Affinity::Upstream)),
                        );
                        self.preedit_range = None;
                    }
                    Handled::Yes
                }
                Ime::Disabled => {
                    if let Some(preedit) = self.preedit_range.clone() {
                        self.text_mut().edit(preedit.clone(), "");
                        self.preedit_range = None;
                        let sm = self.selection.map(|x| x.min()).unwrap_or(0);
                        if preedit.contains(&sm) {
                            self.selection =
                                Some(Selection::caret(preedit.start, Affinity::Upstream));
                        }
                    }
                    Handled::Yes
                }
            },
            TextEvent::ModifierChange(_) => Handled::No,
            TextEvent::FocusChange(_) => Handled::No,
        }
    }
}

impl<T: EditableText> Deref for TextEditor<T> {
    type Target = TextWithSelection<T>;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

// TODO: Being able to call `Self::Target::rebuild` (and `draw`) isn't great.
impl<T: EditableText> DerefMut for TextEditor<T> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.inner
    }
}

#[cfg(test)]
mod tests {
    use super::EditableText;

    // #[test]
    // fn arcstring_empty_edit() {
    //     let a = Arc::new("hello".to_owned());
    //     let mut b = a.clone();
    //     b.edit(5..5, "");
    //     assert!(Arc::ptr_eq(&a, &b));
    // }

    #[test]
    fn replace() {
        let mut a = String::from("hello world");
        a.edit(1..9, "era");
        assert_eq!("herald", a);
    }
}
