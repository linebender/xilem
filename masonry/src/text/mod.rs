// Copyright 2018 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

//! Support for text display and rendering
//!
//! There are three kinds of text commonly needed:
//!  1) Entirely display text (e.g. a button)
//!  2) Selectable text (e.g. a paragraph of content)
//!  3) Editable text (e.g. a search bar)
//!
//! All of these have the same set of global styling options, and can contain rich text

mod store;
pub use store::{Link, TextStorage};

mod layout;
pub use layout::{LayoutMetrics, TextBrush, TextLayout};

mod selection;
pub use selection::{
    len_utf8_from_first_byte, EditableTextCursor, Selectable, StringCursor, TextWithSelection,
};

// mod movement;

mod edit;
pub use edit::{EditableText, TextEditor};

mod backspace;
pub use backspace::offset_for_delete_backwards;
