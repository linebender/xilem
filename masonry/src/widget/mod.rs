// Copyright 2018 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

//! Common widgets.

#[allow(clippy::module_inception)]
pub(crate) mod widget;
mod widget_mut;
mod widget_pod;
mod widget_ref;
mod widget_state;

#[cfg(test)]
mod tests;

mod align;
mod button;
mod checkbox;
mod flex;
mod grid;
mod image;
mod label;
mod portal;
mod progress_bar;
mod prose;
mod root_widget;
mod scroll_bar;
mod sized_box;
mod spinner;
mod split;
mod textbox;
mod variable_label;
mod widget_arena;
mod taffy_layout;

pub use self::image::Image;
pub use align::Align;
pub use button::Button;
pub use checkbox::Checkbox;
pub use flex::{CrossAxisAlignment, Flex, FlexParams, MainAxisAlignment};
pub use grid::{Grid, GridParams};
pub use taffy_layout::TaffyLayout;
pub use label::{Label, LineBreaking};
pub use portal::Portal;
pub use progress_bar::ProgressBar;
pub use prose::Prose;
pub use root_widget::RootWidget;
pub use scroll_bar::ScrollBar;
pub use sized_box::SizedBox;
pub use spinner::Spinner;
pub use split::Split;
pub use textbox::Textbox;
pub use variable_label::VariableLabel;
pub use widget_mut::WidgetMut;
pub use widget_pod::WidgetPod;
pub use widget_ref::WidgetRef;
pub use widget::Axis;

pub(crate) use widget_arena::WidgetArena;
pub(crate) use widget_state::WidgetState;

use crate::{Affine, Size};

// These are based on https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit
/// Strategies for inscribing a rectangle inside another rectangle.
#[derive(Clone, Copy, Default, PartialEq)]
pub enum ObjectFit {
    /// As large as possible without changing aspect ratio of image and all of image shown
    #[default]
    Contain,
    /// As large as possible with no dead space so that some of the image may be clipped
    Cover,
    /// Fill the widget with no dead space, aspect ratio of widget is used
    Fill,
    /// Fill the height with the images aspect ratio, some of the image may be clipped
    FitHeight,
    /// Fill the width with the images aspect ratio, some of the image may be clipped
    FitWidth,
    /// Do not scale
    None,
    /// Scale down to fit but do not scale up
    ScaleDown,
}

// TODO - Need to write tests for this, in a way that's relatively easy to visualize.

impl ObjectFit {
    /// Calculate an origin and scale for an image with a given `ObjectFit`.
    ///
    /// This takes some properties of a widget and an object fit and returns an affine matrix
    /// used to position and scale the image in the widget.
    pub fn affine_to_fill(self, parent: Size, fit_box: Size) -> Affine {
        let raw_scalex = parent.width / fit_box.width;
        let raw_scaley = parent.height / fit_box.height;

        let (scalex, scaley) = match self {
            ObjectFit::Contain => {
                let scale = raw_scalex.min(raw_scaley);
                (scale, scale)
            }
            ObjectFit::Cover => {
                let scale = raw_scalex.max(raw_scaley);
                (scale, scale)
            }
            ObjectFit::Fill => (raw_scalex, raw_scaley),
            ObjectFit::FitHeight => (raw_scaley, raw_scaley),
            ObjectFit::FitWidth => (raw_scalex, raw_scalex),
            ObjectFit::ScaleDown => {
                let scale = raw_scalex.min(raw_scaley).min(1.0);
                (scale, scale)
            }
            ObjectFit::None => (1.0, 1.0),
        };

        let origin_x = (parent.width - (fit_box.width * scalex)) / 2.0;
        let origin_y = (parent.height - (fit_box.height * scaley)) / 2.0;

        Affine::new([scalex, 0., 0., scaley, origin_x, origin_y])
    }
}
