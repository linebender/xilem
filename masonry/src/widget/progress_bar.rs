// Copyright 2019 the Xilem Authors and the Druid Authors
// SPDX-License-Identifier: Apache-2.0

//! A progress bar widget.

use accesskit::Role;
use kurbo::Point;
use smallvec::{smallvec, SmallVec};
use tracing::{trace, trace_span, Span};
use vello::Scene;

use crate::kurbo::Size;
use crate::paint_scene_helpers::{fill_lin_gradient, stroke, UnitPoint};
use crate::text::TextLayout;
use crate::widget::WidgetMut;
use crate::{
    theme, AccessCtx, AccessEvent, ArcStr, BoxConstraints, EventCtx, LayoutCtx, LifeCycle,
    LifeCycleCtx, PaintCtx, PointerEvent, StatusChange, TextEvent, Widget, WidgetId,
};

/// A progress bar
pub struct ProgressBar {
    /// A value in the range `[0, 1]` inclusive, where 0 is 0% and 1 is 100% complete.
    ///
    /// `None` variant can be used to show a progress bar without a percentage.
    /// It is also used if an invalid float (outside of [0, 1]) is passed.
    progress: Option<f32>,
    label: TextLayout<ArcStr>,
}

impl ProgressBar {
    /// Create a new `ProgressBar`.
    ///
    /// `progress` is a number between 0 and 1 inclusive. If it is `NaN`, then an
    /// indefinite progress bar will be shown.
    /// Otherwise, the input will be clamped to [0, 1].
    pub fn new(progress: Option<f32>) -> Self {
        let mut out = Self::new_indefinite();
        out.set_progress(progress);
        out
    }

    fn new_indefinite() -> Self {
        Self {
            progress: None,
            label: TextLayout::new("".into(), crate::theme::TEXT_SIZE_NORMAL as f32),
        }
    }

    fn set_progress(&mut self, mut progress: Option<f32>) {
        clamp_progress(&mut progress);
        // check to see if we can avoid doing work
        if self.progress != progress {
            self.progress = progress;
            self.update_text();
        }
    }

    /// Updates the text layout with the current part-complete value
    fn update_text(&mut self) {
        self.label.set_text(self.value().into());
    }

    fn value(&self) -> ArcStr {
        if let Some(value) = self.progress {
            format!("{:.0}%", value * 100.).into()
        } else {
            "".into()
        }
    }

    fn value_accessibility(&self) -> Box<str> {
        if let Some(value) = self.progress {
            format!("{:.0}%", value * 100.).into()
        } else {
            "progress unspecified".into()
        }
    }
}

// --- MARK: WIDGETMUT ---
impl WidgetMut<'_, ProgressBar> {
    pub fn set_progress(&mut self, progress: Option<f32>) {
        self.widget.set_progress(progress);
        self.ctx.request_layout();
        self.ctx.request_accessibility_update();
    }
}

/// Helper to ensure progress is either a number between [0, 1] inclusive, or `None`.
///
/// NaNs are converted to `None`.
fn clamp_progress(progress: &mut Option<f32>) {
    if let Some(value) = progress {
        if value.is_nan() {
            *progress = None;
        } else {
            *progress = Some(value.clamp(0., 1.));
        }
    }
}

// --- MARK: IMPL WIDGET ---
impl Widget for ProgressBar {
    fn on_pointer_event(&mut self, _ctx: &mut EventCtx, event: &PointerEvent) {
        match event {
            // pointer events unhandled for now
            _ => (),
        }
    }

    fn on_text_event(&mut self, _ctx: &mut EventCtx, _event: &TextEvent) {}

    fn on_access_event(&mut self, ctx: &mut EventCtx, event: &AccessEvent) {
        if event.target == ctx.widget_id() {
            match event.action {
                // access events unhandled for now
                _ => {}
            }
        }
    }

    fn on_status_change(&mut self, ctx: &mut LifeCycleCtx, _event: &StatusChange) {
        ctx.request_paint();
    }

    fn lifecycle(&mut self, _ctx: &mut LifeCycleCtx, _event: &LifeCycle) {}

    fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints) -> Size {
        const DEFAULT_WIDTH: f64 = 400.;

        if self.label.needs_rebuild() {
            let (font_ctx, layout_ctx) = ctx.text_contexts();
            self.label.rebuild(font_ctx, layout_ctx);
        }
        let label_size = self.label.size();

        let desired_size = Size::new(
            DEFAULT_WIDTH.max(label_size.width),
            crate::theme::BASIC_WIDGET_HEIGHT.max(label_size.height),
        );
        let our_size = bc.constrain(desired_size);
        trace!("Computed layout: size={}", our_size);
        our_size
    }

    fn paint(&mut self, ctx: &mut PaintCtx, scene: &mut Scene) {
        let border_width = 1.;

        if self.label.needs_rebuild() {
            debug_panic!("Called ProgressBar paint before layout");
        }

        let rect = ctx
            .size()
            .to_rect()
            .inset(-border_width / 2.)
            .to_rounded_rect(2.);

        fill_lin_gradient(
            scene,
            &rect,
            [theme::BACKGROUND_LIGHT, theme::BACKGROUND_DARK],
            UnitPoint::TOP,
            UnitPoint::BOTTOM,
        );

        stroke(scene, &rect, theme::BORDER_DARK, border_width);

        let progress_rect_size = Size::new(
            ctx.size().width * self.progress.unwrap_or(1.) as f64,
            ctx.size().height,
        );
        let progress_rect = progress_rect_size
            .to_rect()
            .inset(-border_width / 2.)
            .to_rounded_rect(2.);

        fill_lin_gradient(
            scene,
            &progress_rect,
            [theme::PRIMARY_LIGHT, theme::PRIMARY_DARK],
            UnitPoint::TOP,
            UnitPoint::BOTTOM,
        );
        stroke(scene, &progress_rect, theme::BORDER_DARK, border_width);

        // center text
        let widget_size = ctx.size();
        let label_size = self.label.size();
        let text_pos = Point::new(
            ((widget_size.width - label_size.width) * 0.5).max(0.),
            ((widget_size.height - label_size.height) * 0.5).max(0.),
        );
        self.label.draw(scene, text_pos);
    }

    fn accessibility_role(&self) -> Role {
        Role::ProgressIndicator
    }

    fn accessibility(&mut self, ctx: &mut AccessCtx) {
        ctx.current_node().set_value(self.value_accessibility());
    }

    fn children_ids(&self) -> SmallVec<[WidgetId; 16]> {
        smallvec![]
    }

    fn make_trace_span(&self) -> Span {
        trace_span!("ProgressBar")
    }

    fn get_debug_text(&self) -> Option<String> {
        Some(self.value_accessibility().into())
    }
}

// --- MARK: TESTS ---
#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;
    use crate::assert_render_snapshot;
    use crate::testing::{widget_ids, TestHarness, TestWidgetExt};

    #[test]
    fn indeterminate_progressbar() {
        let [progressbar_id] = widget_ids();
        let widget = ProgressBar::new(None).with_id(progressbar_id);

        let mut harness = TestHarness::create(widget);

        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "indeterminate_progressbar");
    }

    #[test]
    fn _0_percent_progressbar() {
        let [_0percent] = widget_ids();

        let widget = ProgressBar::new(Some(0.)).with_id(_0percent);
        let mut harness = TestHarness::create(widget);
        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "0_percent_progressbar");
    }

    #[test]
    fn _25_percent_progressbar() {
        let [_25percent] = widget_ids();

        let widget = ProgressBar::new(Some(0.25)).with_id(_25percent);
        let mut harness = TestHarness::create(widget);
        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "25_percent_progressbar");
    }

    #[test]
    fn _50_percent_progressbar() {
        let [_50percent] = widget_ids();

        let widget = ProgressBar::new(Some(0.5)).with_id(_50percent);
        let mut harness = TestHarness::create(widget);
        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "50_percent_progressbar");
    }

    #[test]
    fn _75_percent_progressbar() {
        let [_75percent] = widget_ids();

        let widget = ProgressBar::new(Some(0.75)).with_id(_75percent);
        let mut harness = TestHarness::create(widget);
        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "75_percent_progressbar");
    }

    #[test]
    fn _100_percent_progressbar() {
        let [_100percent] = widget_ids();

        let widget = ProgressBar::new(Some(1.)).with_id(_100percent);
        let mut harness = TestHarness::create(widget);
        assert_debug_snapshot!(harness.root_widget());
        assert_render_snapshot!(harness, "100_percent_progressbar");
    }
}
