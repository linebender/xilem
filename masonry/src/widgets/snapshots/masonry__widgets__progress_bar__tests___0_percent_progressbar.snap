---
source: masonry/src/widget/progress_bar.rs
expression: harness.root_widget()
snapshot_kind: text
---
SizedBox(
    ProgressBar<0%>(
        Label<0%>,
    ),
)
